File: src/cs/cellular-ga/CellularGrid.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Text;
using aMaze_ingSolver;
using System.Threading.Tasks;
using System.Diagnostics;
using Troschuetz.Random.Generators;
using System.Threading;

namespace cellular_ga
{
    class CellularGrid
    {
        static AbstractGenerator randomGenerator = new MT19937Generator();

        /// <summary>
        /// Number of columns.
        /// </summary>
        public int ColCount { get; }

        /// <summary>
        /// Number of rows.
        /// </summary>
        public int RowCount { get; }

        private Cell[] _currentPopulation;

        public ReplaceType ReplaceMethod { get; set; } = ReplaceType.ReplaceAll;
        public NeighborhoodType NeighborhoodSelectionMethod { get; set; } = NeighborhoodType.L5;


        public CellularGrid(int dimension)
        {
            ColCount = dimension;
            RowCount = dimension;
        }

        public CellularGrid(int rowCount, int colCount)
        {
            ColCount = colCount;
            RowCount = rowCount;
        }

        private Cell GetAt(int row, int col)
        {
            return _currentPopulation[(row * ColCount) + col];
        }

        private Cell GetAtGrid(Cell[] grid, int row, int col)
        {
            return grid[(row * ColCount) + col];
        }

        private void SetAtGrid(Cell[] grid, int row, int col, Cell cell)
        {
            grid[(row * ColCount) + col] = cell;
        }

        private void SetAt(int row, int col, Cell cell)
        {
            _currentPopulation[(row * ColCount) + col] = cell;
        }

        public double GetScoreOfGeneration()
        {
            double totalSum = _currentPopulation.Sum(c => c.GetFitness());
            double result = (totalSum / (double)_currentPopulation.Length) / Cell.MAX_FITNESS_VALUE;
            return result;
        }

        public void Initialize()
        {
            _currentPopulation = new Cell[RowCount * ColCount];

            //for (int row = 0; row < RowCount; row++)
            //{
            //    for (int col = 0; col < ColCount; col++)
            //    {
            //        SetAt(row, col, Cell.GenerateRandomCell(col, row, (byte)((row * col) % byte.MaxValue)));
            //    }
            //}
            byte r, g, b;
            int borderSize = RowCount / 5;
            Random rand = new Random();
            Cell newCell;
            for (int row = 0; row < RowCount; row++)
            {
                for (int col = 0; col < ColCount; col++)
                {
                    r = (byte)rand.Next(0, 256);
                    g = (byte)rand.Next(0, 256);
                    b = (byte)rand.Next(0, 256);

                    if ((row < borderSize) && (col < borderSize))
                    {
                        newCell = new Cell(new Point(col, row)) { R = r, G = g, B = b };
                    }
                    else
                    {
                        newCell = new Cell(new Point(col, row)) { R = 1, G = 1, B = 1 };
                    }
                    SetAt(row, col, newCell);
                }
            }
        }

        public void Evolve(int maxGenerationCount, bool asynchronous, bool saveImagesInDebug = false, string folder = null)
        {
            string format = "Generation: {0}; Score: {1}; Iteration time: {2} ms.";

            double generationScore = GetScoreOfGeneration();
            Console.WriteLine("Chosen neighborhood: {0}\nChoose replace method: {1}", NeighborhoodSelectionMethod.ToString(), ReplaceMethod.ToString());
            Console.WriteLine("Initial generation score: {0}", generationScore);

            Stopwatch s = new Stopwatch();
            for (int generation = 1; generation <= maxGenerationCount; generation++)
            {
                s.Start();
                if (asynchronous)
                    AsynchronousEvolutionStep();
                else
                    SynchronousEvolutionStep();
                s.Stop();

                generationScore = GetScoreOfGeneration();

                Console.WriteLine(format, generation, generationScore, s.Elapsed.TotalMilliseconds);
                s.Reset();


                if (saveImagesInDebug)
                    DebugDumpGridToImage($"{folder}/population{(generation + 1)}.png", System.Drawing.Imaging.ImageFormat.Png);

                if (generationScore == 1.0)
                {
                    Console.WriteLine("Reached our objective.");
                    break;
                }
            }
        }

        public void SynchronousEvolutionStep()
        {
            Cell[] newPopulation = new Cell[RowCount * ColCount];

            Cell cell;
            IEnumerable<Cell> neighborhood;
            Tuple<Cell, Cell> parents;
            Cell offspring;
            for (int row = 0; row < RowCount; row++)
            {
                for (int col = 0; col < ColCount; col++)
                {
                    cell = GetAt(row, col);
                    neighborhood = GetNeighborhood(row, col, NeighborhoodSelectionMethod);

                    parents = Operators.SelectParents(neighborhood);
                    offspring = Operators.Reproduction(col, row, parents);

                    if (ReplaceMethod == ReplaceType.ReplaceWorstInNeighbourhood)
                    {
                        offspring.ToReplace = neighborhood.First(n => n.GetFitness() == neighborhood.Min(x => x.GetFitness())).Location;
                    }
                    else if (ReplaceMethod == ReplaceType.ReplaceOneParent)
                    {
                        offspring.ToReplace = randomGenerator.NextBoolean() ? parents.Item1.Location : parents.Item2.Location;
                    }

                    SetAtGrid(newPopulation, row, col, offspring);
                }
            }

            _currentPopulation = Operators.Replace(RowCount, ColCount, _currentPopulation, newPopulation, ReplaceMethod);
        }

        public async Task MultiThreadedEvolve(int maxGenerationCount, bool asynchronous = false, int? threadCount = null)
        {
            string format = "Generation: {0}; Score: {1}; Iteration time: {2} ms.";

            double generationScore = GetScoreOfGeneration();
            Console.WriteLine("Chosen neighborhood: {0}\nChoose replace method: {1}", NeighborhoodSelectionMethod.ToString(), ReplaceMethod.ToString());
            Console.WriteLine("Initial generation score: {0}", generationScore);

            Stopwatch s = new Stopwatch();
            for (int generation = 1; generation <= maxGenerationCount; generation++)
            {
                s.Start();

                if (asynchronous)
                    await MultithreadedAsynchronousEvolutionStep(threadCount);
                else
                    await MultiThreadEvolutionStep(threadCount);

                s.Stop();

                //for (int i = 0; i < _currentPopulation.Length; i++)
                //{
                //    if (_currentPopulation[i] == null)
                //    {
                //        Console.WriteLine($"{i} is null wtf");
                //    }
                //}

                generationScore = GetScoreOfGeneration();

                Console.WriteLine(format, generation, generationScore, s.Elapsed.TotalMilliseconds);
                s.Reset();

                if (generationScore == 1.0)
                {
                    Console.WriteLine("Reached our objective.");
                }
            }
        }

        private Cell[] WorkerSynchronousEvolutionStep(int fromRow, int toRow)
        {
            AbstractGenerator randomGenerator = new MT19937Generator();
            Cell cell;
            IEnumerable<Cell> neighborhood;
            Tuple<Cell, Cell> parents;
            Cell offspring;

            Cell[] newPopulation = new Cell[(toRow - fromRow) * ColCount];

            int index = 0;
            for (int row = fromRow; row < toRow; row++)
            {
                for (int col = 0; col < ColCount; col++)
                {
                    cell = GetAt(row, col);
                    neighborhood = GetNeighborhood(row, col, NeighborhoodSelectionMethod);

                    parents = Operators.SelectParents(neighborhood);
                    offspring = Operators.Reproduction(col, row, parents, randomGenerator);

                    if (ReplaceMethod == ReplaceType.ReplaceWorstInNeighbourhood)
                    {
                        offspring.ToReplace = neighborhood.First(n => n.GetFitness() == neighborhood.Min(x => x.GetFitness())).Location;
                    }
                    else if (ReplaceMethod == ReplaceType.ReplaceOneParent)
                    {
                        offspring.ToReplace = randomGenerator.NextBoolean() ? parents.Item1.Location : parents.Item2.Location;
                    }

                    newPopulation[index++] = offspring;
                }
            }
            return newPopulation;
        }

        public async Task MultiThreadEvolutionStep(int? suggesterThreadCount = null)
        {
            int workerCount = GetThreadCount(suggesterThreadCount);
            int workerRowCount = RowCount / workerCount;

            Task<Cell[]>[] workers = new Task<Cell[]>[workerCount];
            for (int workerId = 0; workerId < workerCount; workerId++)
            {
                int workerRowFrom = workerId * workerRowCount;
                int workerRowTo = (workerId == workerCount - 1) ? RowCount : workerRowFrom + workerRowCount;

                workers[workerId] = Task<Cell[]>.Factory.StartNew(() => WorkerSynchronousEvolutionStep(workerRowFrom, workerRowTo));
            }

            var taskResult = await Task.WhenAll(workers);

            Cell[] newPopulation = new Cell[RowCount * ColCount];

            int offset = 0;

            for (int workerId = 0; workerId < workerCount; workerId++)
            {
                int copyLen = taskResult[workerId].Length;
                Array.ConstrainedCopy(taskResult[workerId], 0, newPopulation, offset, copyLen);
                offset += copyLen;
            }
            _currentPopulation = Operators.Replace(RowCount, ColCount, _currentPopulation, newPopulation, ReplaceMethod);
        }

        public void DebugDumpGridToImage(string filename, System.Drawing.Imaging.ImageFormat format)
        {
#if DEBUG
            using (Bitmap gridBitmap = new Bitmap(ColCount, RowCount))
            {
                using (BitmapPlus bmpP = new BitmapPlus(gridBitmap, System.Drawing.Imaging.ImageLockMode.ReadWrite))
                {
                    Cell cell;
                    for (int row = 0; row < RowCount; row++)
                    {
                        for (int col = 0; col < ColCount; col++)
                        {
                            cell = GetAt(row, col);
                            bmpP.SetPixel(col, row, Color.FromArgb(cell.R, cell.G, cell.B));
                        }
                    }
                }

                gridBitmap.Save(filename, format);
            } 
#endif
        }

        public List<Cell> GetNeighborhood(int row, int col, NeighborhoodType type)
        {
            // There 
            int MyMod(int x, int mod) => (x >= 0) ? (x % mod) : (x + mod);

            List<Cell> neighborhood = new List<Cell>();


            switch (type)
            {
                case NeighborhoodType.L5:
                    {
                        neighborhood.Add(GetAt(row, col));

                        neighborhood.Add(GetAt(row, MyMod(col - 1, ColCount))); // Left
                        neighborhood.Add(GetAt(MyMod(row - 1, RowCount), col)); // Top
                        neighborhood.Add(GetAt(row, MyMod(col + 1, ColCount))); // Right
                        neighborhood.Add(GetAt(MyMod(row + 1, RowCount), col)); // Bottom

                        Debug.Assert(neighborhood.Count == 5);
                    }
                    break;
                case NeighborhoodType.L9:
                    {
                        neighborhood.Add(GetAt(row, col));

                        neighborhood.Add(GetAt(row, MyMod(col - 1, ColCount))); // Left
                        neighborhood.Add(GetAt(row, MyMod(col - 2, ColCount))); // Left 2
                        neighborhood.Add(GetAt(MyMod(row - 1, RowCount), col)); // Top
                        neighborhood.Add(GetAt(MyMod(row - 2, RowCount), col)); // Top 2
                        neighborhood.Add(GetAt(row, MyMod(col + 1, ColCount))); // Right
                        neighborhood.Add(GetAt(row, MyMod(col + 2, ColCount))); // Right 2
                        neighborhood.Add(GetAt(MyMod(row + 1, RowCount), col)); // Bottom
                        neighborhood.Add(GetAt(MyMod(row + 2, RowCount), col)); // Bottom 2

                        Debug.Assert(neighborhood.Count == 9);
                    }
                    break;
                case NeighborhoodType.C9:
                    {
                        for (int nRow = row - 1; nRow <= row + 1; nRow++)
                        {
                            for (int nCol = col - 1; nCol <= col + 1; nCol++)
                            {

                                neighborhood.Add(GetAt(MyMod(nRow, RowCount), MyMod(nCol, ColCount)));
                            }
                        }
                        Debug.Assert(neighborhood.Count == 9);

                    }
                    break;
                case NeighborhoodType.C13:
                    {
                        for (int nRow = row - 1; nRow <= row + 1; nRow++)
                        {
                            for (int nCol = col - 1; nCol <= col + 1; nCol++)
                            {
                                neighborhood.Add(GetAt(MyMod(nRow, RowCount), MyMod(nCol, ColCount)));
                            }
                        }
                        neighborhood.Add(GetAt(row, MyMod(col - 2, ColCount))); // Left 2
                        neighborhood.Add(GetAt(MyMod(row - 2, RowCount), col)); // Top 2
                        neighborhood.Add(GetAt(row, MyMod(col + 2, ColCount))); // Right 2
                        neighborhood.Add(GetAt(MyMod(row + 2, RowCount), col)); // Bottom 2

                        Debug.Assert(neighborhood.Count == 13);
                    }
                    break;
                default:
                    throw new Exception("INVALID TYPE.");
            }

            return neighborhood;
        }

        public void AsynchronousEvolutionStep()
        {
            Cell[] newPopulation = new Cell[RowCount * ColCount];

            Cell cell;
            IEnumerable<Cell> neighborhood;
            Tuple<Cell, Cell> parents;
            Cell offspring;
            for (int row = 0; row < RowCount; row++)
            {
                for (int col = 0; col < ColCount; col++)
                {
                    cell = GetAt(row, col);
                    neighborhood = GetNeighborhood(row, col, NeighborhoodSelectionMethod);

                    parents = Operators.SelectParents(neighborhood);
                    offspring = Operators.Reproduction(col, row, parents);

                    if (ReplaceMethod == ReplaceType.ReplaceWorstInNeighbourhood)
                    {
                        offspring.ToReplace = neighborhood.First(n => n.GetFitness() == neighborhood.Min(x => x.GetFitness())).Location;
                    }
                    else if (ReplaceMethod == ReplaceType.ReplaceOneParent)
                    {
                        offspring.ToReplace = randomGenerator.NextBoolean() ? parents.Item1.Location : parents.Item2.Location;
                    }

                    SetAtGrid(newPopulation, row, col, offspring);
                }
                _currentPopulation = Operators.ReplaceRow(row, ColCount, _currentPopulation, newPopulation, ReplaceMethod);
            }
        }

        private int GetThreadCount(int? suggested)
        {
            int result = suggested.HasValue ? suggested.Value : Environment.ProcessorCount;
            return result;
        }

        public async Task MultithreadedAsynchronousEvolutionStep(int? suggestedThreadCount = null)
        {
            int workerCount = GetThreadCount(suggestedThreadCount);
            int workerColCount = ColCount / workerCount;

            Cell[] newPopulation = new Cell[RowCount * ColCount];

            Cell[] newRow;
            
            int offset, copyLen;

            Task<Cell[]>[] workers = new Task<Cell[]>[workerCount];
            for (int row = 0; row < RowCount; row++)
            {
                for (int workerId = 0; workerId < workerCount; workerId++)
                {
                    int workerColFrom = workerId * workerColCount;
                    int workerColTo = (workerId == workerCount - 1) ? ColCount : workerColFrom + workerColCount;


                    workers[workerId] = Task<Cell[]>.Factory.StartNew(() => WorkerAsynchronousEvolutionStep(row, workerColFrom, workerColTo));
                }

                newRow = new Cell[ColCount];

                var taskResult = await Task.WhenAll(workers);


                offset = 0;
                for (int workerId = 0; workerId < workerCount; workerId++)
                {
                    copyLen = taskResult[workerId].Length;
                    Array.ConstrainedCopy(taskResult[workerId], 0, newRow, offset, copyLen);
                    offset += copyLen;
                }

                _currentPopulation = Operators.SpecialReplaceRow(row, ColCount, _currentPopulation, newRow, ReplaceMethod);
            }
        }

        private Cell[] WorkerAsynchronousEvolutionStep(int row, int colFrom, int colTo)
        {
            AbstractGenerator randomGenerator = new MT19937Generator();
            Cell cell;
            IEnumerable<Cell> neighborhood;
            Tuple<Cell, Cell> parents;
            Cell offspring;

            int index = 0;
            Cell[] result = new Cell[colTo - colFrom];

            for (int col = colFrom; col < colTo; col++)
            {
                cell = GetAt(row, col);
                neighborhood = GetNeighborhood(row, col, NeighborhoodSelectionMethod);

                parents = Operators.SelectParents(neighborhood);
                offspring = Operators.Reproduction(col, row, parents, randomGenerator);

                if (ReplaceMethod == ReplaceType.ReplaceWorstInNeighbourhood)
                {
                    offspring.ToReplace = neighborhood.First(n => n.GetFitness() == neighborhood.Min(x => x.GetFitness())).Location;
                }
                else if (ReplaceMethod == ReplaceType.ReplaceOneParent)
                {
                    offspring.ToReplace = randomGenerator.NextBoolean() ? parents.Item1.Location : parents.Item2.Location;
                }
                result[index++] = offspring;
            }
            return result;
        }
    }
}

File: src/cpp/cellular_grid.cpp
#include "cellular_grid.h"
#include "image.cpp"
#include <omp.h>

Cell &CellularGrid::at(uint row, uint col)
{
    //std::lock_guard<std::mutex> lock(currentPopulationMutex);
    return currentPopulation[(row * colCount) + col];
}

CellularGrid::CellularGrid(const uint dimension)
{
    colCount = dimension;
    rowCount = dimension;
}
CellularGrid::CellularGrid(const uint width, const uint height)
{
    rowCount = height;
    colCount = width;
}
CellularGrid::~CellularGrid()
{
    currentPopulation.clear();
    currentPopulation.shrink_to_fit();
}

void CellularGrid::initialize(const NeighborhoodType neighborhoodType, const PopulationMergeType mergeMethod, const InitializationType initType)
{
    this->neighborhoodMethod = neighborhoodType;
    this->mergeMethod = mergeMethod;
    currentPopulation = std::vector<Cell>();
    currentPopulation.resize(rowCount * colCount);

    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<int> dis(0, 255); // Both ranges are inclusive.

    uchar r, g, b, discrimination;
    Cell newCell;
    uint index;
    switch (initType)
    {
    case RandomWithDiscrimination:
    {
#pragma omp parallel for private(r, g, b, discrimination)
        for (uint row = 0; row < rowCount; row++)
        {
            for (uint col = 0; col < colCount; col++)
            {
                discrimination = (uchar)((row * col) % UCHAR_MAX_AS_INT);
                r = (uchar)dis(gen);
                g = (uchar)dis(gen);
                b = (uchar)dis(gen);

                r = (r > discrimination) ? (uchar)(r - discrimination) : r;
                g = (g > discrimination) ? (uchar)(g - discrimination) : g;
                b = (b > discrimination) ? (uchar)(b - discrimination) : b;

#pragma omp critical
                {
                    currentPopulation[(row * colCount) + col] = Cell(Point(col, row), r, g, b);
                }
            }
        }
    }
    break;
    case FitBorders:
    {
        uint borderSize = rowCount / 15;

#pragma omp parallel for private(r, g, b, newCell)
        for (uint row = 0; row < rowCount; row++)
        {
            for (uint col = 0; col < colCount; col++)
            {
                r = (uchar)dis(gen);
                g = (uchar)dis(gen);
                b = (uchar)dis(gen);

                if ((row < borderSize) || (col < borderSize) || (row > (rowCount - borderSize)) || (col > (colCount - borderSize)))
                {
                    newCell = Cell(Point(col, row), r, g, b);
                }
                else
                {
                    newCell = Cell(Point(col, row), 1, 1, 1);
                }

#pragma omp critical
                {
                    currentPopulation[(row * colCount) + col] = newCell;
                }
            }
        }
    }
    break;
    case FitCorner:
    {
        uint borderSize = rowCount / 10;

#pragma omp parallel for private(r, g, b, newCell)
        for (uint row = 0; row < rowCount; row++)
        {
            for (uint col = 0; col < colCount; col++)
            {
                r = (uchar)dis(gen);
                g = (uchar)dis(gen);
                b = (uchar)dis(gen);

                if ((row < borderSize) && (col < borderSize))
                {
                    newCell = Cell(Point(col, row), r, g, b);
                }
                else
                {
                    newCell = Cell(Point(col, row), 1, 1, 1);
                }
#pragma omp critical
                {
                    currentPopulation[(row * colCount) + col] = newCell;
                }
            }
        }
    }
    break;
    default:
        assert(false && "Wrong initialization type.");
    }
}

double CellularGrid::get_score_of_generation() const
{
    double sum = 0.0;

    for (size_t i = 0; i < currentPopulation.size(); i++)
    {
        assert(currentPopulation[i].get_fitness() <= MAX_FITNESS_VALUE);
        sum += currentPopulation[i].get_fitness();
    }
    double result = (sum / (double)currentPopulation.size()) / MAX_FITNESS_VALUE;
    return result;
}

void CellularGrid::dump_current_population_to_image(const std::string &folder, uint generation, bool bw)
{
    cimg_library::CImg<uchar> gridImage;
    if (bw)
        gridImage = create_grayscale_image(colCount, rowCount);
    else
        gridImage = create_color_image(colCount, rowCount);

    for (uint row = 0; row < rowCount; row++)
    {
        for (uint col = 0; col < colCount; col++)
        {
            Cell cell = at(row, col);
            if (bw)
            {
                uchar normalized = (uchar)((cell.get_fitness() / MAX_FITNESS_VALUE) * 255.0f);
                set_pixel(gridImage, row, col, normalized);
            }
            else
            {
                set_pixel(gridImage, row, col, RgbPixel(cell.R, cell.G, cell.B));
            }
        }
    }

    std::string filename = folder + "/generation_" + std::to_string(generation) + ".bmp";
    gridImage.save_bmp(filename.c_str());
}

void CellularGrid::evolve(const int maxGenerationCount, const bool multiThreaded, const int threadCount, const bool saveImages, const std::string &folder)
{
    double generationScore = get_score_of_generation();
    printf("Chosen neighborhood: %s\nChosen merge method: %s\n", std::to_string(neighborhoodMethod).c_str(), std::to_string(mergeMethod).c_str());
    printf("Initial generation score: %f\n", generationScore);

    StopwatchData s;
    double time;
    for (int generation = 1; generation <= maxGenerationCount; generation++)
    {
        start_stopwatch(s);
        if (multiThreaded)
            openmp_evolution_step(threadCount); //multithreaded_evolution_step(threadCount);
        else
            synchronous_evolution_step();
        stop_stopwatch(s);

        time = elapsed_milliseconds(s);
        generationScore = get_score_of_generation();

        printf("Completed generation %i; Score: %f; Iteration time %f ms\n", generation, generationScore, time);

        if (saveImages)
            dump_current_population_to_image(folder, generation, true);

        if (generationScore == 1.0)
        {
            printf("Target objective was reached.\n");
            return;
        }
    }
}

void CellularGrid::synchronous_evolution_step()
{
    std::vector<Cell> newPopulation;
    newPopulation.resize(rowCount * colCount);

    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<int> dis012(0, 2); // Both ranges are inclusive.
    std::uniform_int_distribution<int> dis01(0, 1);  // Both ranges are inclusive.
    int random;
    for (uint row = 0; row < rowCount; row++)
    {
        for (uint col = 0; col < colCount; col++)
        {
            std::vector<Cell> neighborhood = get_neighborhood(row, col);
            std::pair<Cell, Cell> parents = select_parents(neighborhood);
            Cell offspring = reproduction(col, row, parents, dis012(gen));

            if (mergeMethod == ReplaceWorstInNeighborhood)
            {
                offspring.cellToReplaceLocation = get_worst_cell(neighborhood).cellLocation;
            }
            else if (mergeMethod == ReplaceOneParent)
            {
                random = dis01(gen);
                offspring.cellToReplaceLocation = (random == 0) ? parents.first.cellLocation : parents.second.cellLocation;
            }
            newPopulation[(row * colCount) + col] = offspring;
        }
    }
    currentPopulation = replace(rowCount, colCount, currentPopulation, newPopulation, mergeMethod);
}

std::vector<Cell> CellularGrid::get_neighborhood(const uint row, const uint col)
{
    std::vector<Cell> neighborhood;
    switch (neighborhoodMethod)
    {
    case L5:
    {
        neighborhood.reserve(5);

        neighborhood.push_back(at(row, col));

        neighborhood.push_back(at(row, mod(col - 1, colCount))); // Left
        neighborhood.push_back(at(mod(row - 1, rowCount), col)); // Top
        neighborhood.push_back(at(row, mod(col + 1, colCount))); // Right
        neighborhood.push_back(at(mod(row + 1, rowCount), col)); // Bottom

        assert(neighborhood.size() == 5);
        return neighborhood;
    }
    break;
    case L9:
    {
        neighborhood.reserve(9);

        neighborhood.push_back(at(row, col));
        neighborhood.push_back(at(row, mod(col - 1, colCount))); // Left
        neighborhood.push_back(at(row, mod(col - 2, colCount))); // Left 2
        neighborhood.push_back(at(mod(row - 1, rowCount), col)); // Top
        neighborhood.push_back(at(mod(row - 2, rowCount), col)); // Top 2
        neighborhood.push_back(at(row, mod(col + 1, colCount))); // Right
        neighborhood.push_back(at(row, mod(col + 2, colCount))); // Right 2
        neighborhood.push_back(at(mod(row + 1, rowCount), col)); // Bottom
        neighborhood.push_back(at(mod(row + 2, rowCount), col)); // Bottom 2

        assert(neighborhood.size() == 9);
        return neighborhood;
    }
    break;
    case C9:
    {
        neighborhood.reserve(9);

        // When this was calculated in for cycles, OpenMp didn't execute them.
        int fromRow = row - 1;
        int toRow = row + 2;
        int fromCol = col - 1;
        int toCol = col + 2;

        for (int nRow = fromRow; nRow < toRow; nRow++)
        {
            for (int nCol = fromCol; nCol < toCol; nCol++)
            {
                neighborhood.push_back(at(mod(nRow, rowCount), mod(nCol, colCount)));
            }
        }

        assert(neighborhood.size() == 9);
        return neighborhood;
    }
    break;
    case C13:
    {
        neighborhood.reserve(13);

        // When this was calculated in for cycles, OpenMp didn't execute them.
        int fromRow = row - 1;
        int toRow = row + 2;
        int fromCol = col - 1;
        int toCol = col + 2;

        for (int nRow = fromRow; nRow < toRow; nRow++)
        {
            for (int nCol = fromCol; nCol < toCol; nCol++)
            {
                neighborhood.push_back(at(mod(nRow, rowCount), mod(nCol, colCount)));
            }
        }
        neighborhood.push_back(at(row, mod(col - 2, colCount))); // Left 2
        neighborhood.push_back(at(mod(row - 2, rowCount), col)); // Top 2
        neighborhood.push_back(at(row, mod(col + 2, colCount))); // Right 2
        neighborhood.push_back(at(mod(row + 2, rowCount), col)); // Bottom 2

        assert(neighborhood.size() == 13);
        return neighborhood;
    }
    break;
    default:
    {
        assert(false && "Wrong method");
    }
    }
}

void CellularGrid::worker_job(int rowFrom, int rowTo, std::vector<Cell> &result)
{
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<int> dis012(0, 2); // Both ranges are inclusive.
    std::uniform_int_distribution<int> dis01(0, 1);  // Both ranges are inclusive.

    Cell offspring;
    std::vector<Cell> neighborhood;
    std::pair<Cell, Cell> parents;

    //  std::vector<Cell> result;

    result.reserve((rowTo - rowFrom) * colCount);
    int index = 0;
    int random;

    for (uint row = rowFrom; row < rowTo; row++)
    {
        for (uint col = 0; col < colCount; col++)
        {
            neighborhood = get_neighborhood(row, col);
            parents = select_parents(neighborhood);
            offspring = reproduction(col, row, parents, dis012(gen));

            if (mergeMethod == ReplaceWorstInNeighborhood)
            {
                offspring.cellToReplaceLocation = get_worst_cell(neighborhood).cellLocation;
            }
            else if (mergeMethod == ReplaceOneParent)
            {
                random = dis01(gen);
                offspring.cellToReplaceLocation = (random == 0) ? parents.first.cellLocation : parents.second.cellLocation;
            }

            result.push_back(offspring);
        }
    }
}

void CellularGrid::multithreaded_evolution_step(const int threadCount)
{
    std::vector<std::vector<Cell>> workersResults;
    workersResults.reserve(threadCount);
    {
        std::vector<std::thread> workers;

        workers.reserve(threadCount);

        int workerRowCount = rowCount / threadCount;

        int workerRowFrom, workerRowTo;
        for (int workerId = 0; workerId < threadCount; workerId++)
        {
            workerRowFrom = workerId * workerRowCount;
            workerRowTo = (workerId == threadCount - 1) ? rowCount : workerRowFrom + workerRowCount;

            workersResults.push_back(std::vector<Cell>());

            workers[workerId] = std::thread(&CellularGrid::worker_job, this, workerRowFrom, workerRowTo, std::ref(workersResults[workerId]));
        }

        std::vector<Cell> newPopulation;
        newPopulation.reserve(rowCount * colCount);

        int offset = 0;
        for (int workerId = 0; workerId < threadCount; workerId++)
        {
            workers[workerId].join();
            //printf("Thread %i reportedly completed.\n", workerId);
            newPopulation.insert((newPopulation.begin() + offset), workersResults[workerId].begin(), workersResults[workerId].end());
            offset += workersResults[workerId].size();
        }

        //std::lock_guard<std::mutex> lock(currentPopulationMutex);
        currentPopulation = replace(rowCount, colCount, currentPopulation, newPopulation, mergeMethod);
    }
}

void print_point(const Point &p)
{
    printf("Cell: at [%i;%i]\n", p.x, p.y);
}
void print_neighborhood(const std::vector<Cell> &neigh)
{
    for (const Cell &n : neigh)
    {
        print_point(n.cellLocation);
    }
}

void CellularGrid::openmp_evolution_step(const int threadCount)
{
    omp_set_num_threads(threadCount);

    std::vector<Cell> newPopulation;
    newPopulation.resize(rowCount * colCount);

    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<int> dis012(0, 2); // Both ranges are inclusive.
    std::uniform_int_distribution<int> dis01(0, 1);  // Both ranges are inclusive.
    int random;

#pragma omp parallel for
    for (uint row = 0; row < rowCount; row++)
    {
        for (uint col = 0; col < colCount; col++)
        {
            std::vector<Cell> neighborhood = get_neighborhood(row, col);
            std::pair<Cell, Cell> parents = select_parents(neighborhood);
            Cell offspring = reproduction(col, row, parents, dis012(gen));

            if (mergeMethod == ReplaceWorstInNeighborhood)
            {
                offspring.cellToReplaceLocation = get_worst_cell(neighborhood).cellLocation;
            }
            else if (mergeMethod == ReplaceOneParent)
            {
                random = dis01(gen);
                offspring.cellToReplaceLocation = (random == 0) ? parents.first.cellLocation : parents.second.cellLocation;
            }

#pragma omp critical
            {
                newPopulation[(row * colCount) + col] = offspring;
            }
        }
    }

    currentPopulation = replace(rowCount, colCount, currentPopulation, newPopulation, mergeMethod);
}
